### PR TITLE
Adding validations to models

### PR DESCRIPTION
**Pull request by larshoubak**
1. Created validations in models: superpower.rb and booking.rb

2. Installed Cloudinary and Carrierwave:

- Setup gems: 'dotenv-rails',  'cloudinary', 'carrierwave'
- Set up new Cloudinary account
- setup new controller: uploader Photo

Must be done:
- test that image upload works
- Test that image display works
- Add this to the superpower controller:
# app/controllers/superpowers_controller.rb
def superpower_params    params.require(:superpower).permit(:name, :price, :description, :photo)
end

### CODE REVIEW

File: app/models/booking.rb [hunk 1]
```diff
@@ -1,4 +1,15 @@
...
   belongs_to :superpower
   belongs_to :rentee, class_name: 'User'
+  validates :start_date, :end_date, presence: true
+  validate :test_end_date_versus_start_date
+
+  private
+
+  def test_end_date_versus_start_date
+    return if end_date.blank? || start_date.blank?
+    if end_date <= start_date
```
Comment (ax-ab) [inline app/models/booking.rb]: should be == and not <= as the user should be allowed to book for one day only, right?

Comment (ax-ab) [inline app/models/booking.rb]: should be == and not <= as the user should be allowed to book for one day only, right?

File: app/models/booking.rb [hunk 2]
```diff
@@ -1,4 +1,15 @@
 class Booking < ApplicationRecord
   belongs_to :superpower
   belongs_to :rentee, class_name: 'User'
+  validates :start_date, :end_date, presence: true
+  validate :test_end_date_versus_start_date
+
+  private
+
+  def test_end_date_versus_start_date
+    return if end_date.blank? || start_date.blank?
```
Comment (ax-ab) [inline app/models/booking.rb]: why check for this when this should be covered in line 4?

File: app/models/superpower.rb [hunk 3]
```diff
@@ -3,6 +3,12 @@ class Superpower < ApplicationRecord
   belongs_to :superpower_category
   has_many :bookings
 
+  validates :name, :description, :image, :price, presence: true
+  validates :name, uniqueness: true
```
Comment (ax-ab) [inline app/models/superpower.rb]: needs to be agreed within the team :) (@fsansalvadore, @fitzsimonsjl). I suggest that we start out by having it non-unique and then do some tests and see how it works and then we can always restrict it later

File: app/models/superpower_category.rb [hunk 4]
```diff
@@ -1,5 +1,3 @@
 class SuperpowerCategory < ApplicationRecord
   has_many :superpowers
-
-
```
Comment (ax-ab) [inline app/models/superpower_category.rb]: should have name presence true and unique, right as we can't have a category twice and with an empty name. same goes for icon.

File: app/models/superpower.rb [hunk 5]
```diff
@@ -3,6 +3,12 @@ class Superpower < ApplicationRecord
...
+
+  mount_uploader :photo, PhotoUploader
```
Comment (ax-ab) [inline app/models/superpower.rb]: should be :image and not :photo to reflect the field in the db schema

File: app/models/superpower.rb [hunk 6]
```diff
@@ -3,6 +3,12 @@ class Superpower < ApplicationRecord
...
+  mount_uploader :tumbnail, PhotoUploader
```
Comment (ax-ab) [inline app/models/superpower.rb]: what is this? if it is a small version of the image we should simply rerender it in the view and not create a seperate field

File: app/models/booking.rb [hunk 7]
```diff
@@ -1,4 +1,15 @@
...
+    if end_date <= start_date
+      errors.add(:end_date, "is after the start date")
```
Comment (ax-ab) [inline app/models/booking.rb]: makes sense but I haven't seen this type of code before. has it been tested?

File: app/models/booking.rb [hunk 8]
```diff
@@ -1,4 +1,15 @@
 class Booking < ApplicationRecord
   belongs_to :superpower
   belongs_to :rentee, class_name: 'User'
+  validates :start_date, :end_date, presence: true
+  validate :test_end_date_versus_start_date
```
Comment (ax-ab) [inline app/models/booking.rb]: 'validates' and not 'validate'?

File: app/models/superpower.rb [hunk 9]
```diff
@@ -3,6 +3,12 @@ class Superpower < ApplicationRecord
   belongs_to :superpower_category
   has_many :bookings
 
+  validates :name, :description, :image, :price, presence: true
```
Comment (ax-ab) [inline app/models/superpower.rb]: please add: validates price must be a positive integer

File: app/models/booking.rb (hunk 1 above)
Comment (larshoubak) [reply app/models/booking.rb]: 11/11 - 12/11 => -1, false

12/11 - 12/11 = 0, false

13/11 - 12/11 = 1, true

File: app/models/superpower.rb (hunk 3 above)
Comment (larshoubak) [reply app/models/superpower.rb]: jep

File: app/models/superpower_category.rb (hunk 4 above)
Comment (larshoubak) [reply app/models/superpower_category.rb]: Good point. I make the change.

File: app/models/booking.rb (hunk 8 above)
Comment (larshoubak) [reply app/models/booking.rb]: I double checked this and found some forums with notes saying that the :validate declaration is used for custom validation (like this method) where as :validates is used for generic validation like presence, uniqueness etc on a field

File: app/models/superpower.rb (hunk 9 above)
Comment (larshoubak) [reply app/models/superpower.rb]: Good point

File: app/models/booking.rb (hunk 7 above)
Comment (larshoubak) [reply app/models/booking.rb]: I wrote the code and asked Irine - she approved it. But yes - must be tested. Proposes that I add this as a task and we test it when the view is ready?

File: app/models/superpower.rb (hunk 6 above)
Comment (larshoubak) [reply app/models/superpower.rb]: Yes, is a small version. I used two image sizes for the cocktail application. 
Fine with approach to rerender in the view.

File: app/models/superpower.rb (hunk 5 above)
Comment (larshoubak) [reply app/models/superpower.rb]: Agree

File: app/models/booking.rb (hunk 1 above)
Comment (ax-ab) [reply app/models/booking.rb]: OK. let' bring it up for group discussion later if needed

File: app/models/booking.rb (hunk 2 above)
Comment (ax-ab) [reply app/models/booking.rb]: Lars will test and delete line if needed

File: app/models/superpower.rb (hunk 6 above)
Comment (ax-ab) [reply app/models/superpower.rb]: lars will delete as agreed

File: app/models/booking.rb (hunk 8 above)
Comment (ax-ab) [reply app/models/booking.rb]: Axel is wrong. We will keep. lars will test

**Pull request merged**